Mathlib/Algebra/DirectSum/Basic.lean

```diff
@@ -394,6 +394,50 @@ theorem finite_support (A : ι → S) (x : DirectSum ι fun i => A i) :
   classical
   exact (DFinsupp.support x).finite_toSet.subset (DirectSum.support_subset _ x)
 
+section map
+
+variable {ι : Type*} [DecidableEq ι] {α : ι → Type*} {β : ι → Type*} [∀ i, AddCommMonoid (α i)]
+variable [∀ i, AddCommMonoid (β i)] (f : ∀(i : ι), α i →+ β i)
+
+/-- create a homomorphism from `⨁ i, α i` to `⨁ i, β i` by giving the component-wise map `f`. -/
+def map : (⨁ i, α i) →+ ⨁ i, β i :=
+  DirectSum.toAddMonoid (fun i : ι ↦ (of β i).comp (f i))
+
+@[simp] lemma map_of (i : ι) (x : α i) : (map f) (of α i x) = of β i (f i x) := by
+  simp [map]
+
+@[simp] lemma map_apply (i : ι) (x : ⨁ i, α i) : (map f) x i = f i (x i) := by
+  apply DirectSum.induction_on (C := fun x ↦ (map f) x i = f i (x i)) x
+  · simp
+  · intro j x; rw [map_of, of_apply, of_apply]
+    by_cases h : j = i
+    · simp [h]; subst h; rfl
+    · simp [h]
+  · simp +contextual
+
+@[simp] lemma map_id : (map (fun i ↦ AddMonoidHom.id (α i))) = AddMonoidHom.id (⨁ i, α i) := by
+  ext i x; simp
+
+@[simp] lemma map_comp {γ : ι → Type*} [∀ i, AddCommMonoid (γ i)] (g : ∀ (i : ι), β i →+ γ i) :
+    (map (fun i ↦ (g i).comp (f i))) = (map g).comp (map f) := by
+  ext i x; simp
+
+lemma map_surjective (h : ∀ i, Function.Surjective (f i)) : Function.Surjective (map f) := by
+  intro x
+  apply DirectSum.induction_on (C := fun x ↦ ∃ a, (map f) a = x) x
+  · use 0; rfl
+  · intro i x; obtain ⟨y, hy⟩ := h i x; use of α i y; simp [hy]
+  · rintro x y ⟨u, hu⟩ ⟨v, hv⟩; use (u + v); simp [hu, hv]
```

Comment:
I realize this is longer, but I think it better reflects Mathlib style. We generally avoid putting several tactics on one line, even if they manage to fit. Sometimes people do two, but almost never 3 or 4. Also, it's nice to use the `induction` tactic, and the `rfl` in the `obtain` / `rcases` pattern syntax.
```suggestion
  induction x using DirectSum.induction_on with
  | H_zero => exact ⟨0, by simp⟩
  | H_basic i x =>
    obtain ⟨y, rfl⟩ := h i x
    exact ⟨of α i y, by simp⟩
  | H_plus x y hx hy => 
    obtain ⟨u, rfl⟩ := hx
    obtain ⟨v, rfl⟩ := hy
    exact ⟨u + v, by simp⟩
```